source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc

```diff
@@ -7,53 +7,168 @@ namespace Extensions {
+                host_to_sessions_.erase(host_sessions_it);
+              }
+            }
+          })) {}
+
+UdpProxyFilter::ClusterInfo::~ClusterInfo() {
+  member_update_cb_handle_->remove();
+  // Sanity check the session accounting. This is not as fast as a straight teardown, but this is
+  // not a performance critical path.
+  while (!sessions_.empty()) {
+    removeSession(sessions_.begin()->get());
+  }
+  ASSERT(host_to_sessions_.empty());
+}
+
+void UdpProxyFilter::ClusterInfo::onData(Network::UdpRecvData& data) {
   const auto active_session_it = sessions_.find(data.addresses_);
   ActiveSession* active_session;
   if (active_session_it == sessions_.end()) {
-    // TODO(mattklein123): Session circuit breaker.
-    // TODO(mattklein123): Instead of looking up the cluster each time, keep track of it via
-    // cluster manager callbacks.
-    Upstream::ThreadLocalCluster* cluster = config_->getCluster();
-    // TODO(mattklein123): Handle the case where the cluster does not exist and add stat.
-    ASSERT(cluster != nullptr);
+    if (!cluster_.info()
+             ->resourceManager(Upstream::ResourcePriority::Default)
+             .connections()
+             .canCreate()) {
+      cluster_.info()->stats().upstream_cx_overflow_.inc();
+      return;
+    }
 
     // TODO(mattklein123): Pass a context and support hash based routing.
-    Upstream::HostConstSharedPtr host = cluster->loadBalancer().chooseHost(nullptr);
-    // TODO(mattklein123): Handle the case where the host does not exist.
-    ASSERT(host != nullptr);
+    Upstream::HostConstSharedPtr host = cluster_.loadBalancer().chooseHost(nullptr);
+    if (host == nullptr) {
+      cluster_.info()->stats().upstream_cx_none_healthy_.inc();
+      return;
+    }
 
-    auto new_session = std::make_unique<ActiveSession>(*this, std::move(data.addresses_), host);
-    active_session = new_session.get();
-    sessions_.emplace(std::move(new_session));
+    active_session = createSession(std::move(data.addresses_), host);
   } else {
-    // TODO(mattklein123): Handle the host going away going away or failing health checks.
     active_session = active_session_it->get();
+    if (active_session->host().health() == Upstream::Host::Health::Unhealthy) {
+      // If a host becomes unhealthy, we optimally would like to replace it with a new session
+      // to a healthy host. We may eventually want to make this behavior configurable, but for now
+      // this will be the universal behavior.
+
+      // TODO(mattklein123): Pass a context and support hash based routing.
+      Upstream::HostConstSharedPtr host = cluster_.loadBalancer().chooseHost(nullptr);
+      if (host != nullptr && host->health() != Upstream::Host::Health::Unhealthy &&
+          host.get() != &active_session->host()) {
+        ENVOY_LOG(debug, "upstream session unhealthy, recreating the session");
+        removeSession(active_session);
+        active_session = createSession(std::move(data.addresses_), host);
```

Comment:
Can we just add an interface to update host_ in the original active_session?

Reply:
I was mostly trying to avoid special case logic, and this should not be a frequent operation. If you feel strongly about it I can change it though. WDYT?